src/hotspot/cpu/x86/icache_x86.cpp

```diff
@@ -27,11 +27,54 @@
 
 #define __ _masm->
 
-void ICacheStubGenerator::generate_icache_flush(ICache::flush_icache_stub_t* flush_icache_stub) {
-  StubCodeMark mark(this, "ICache", "flush_icache_stub");
+void x86_generate_icache_fence(MacroAssembler* _masm) {
+  switch (X86ICacheSync) {
+    case 0:
+      break;
+    case 1:
+      __ mfence();
+      break;
+    case 2:
+    case 3:
+      __ sfence();
+      break;
+    case 4:
+      __ push(rax);
```

Comment:
x86 also has `serialize` which does, you guess what, serialize the instruction stream. I suggest adding a routine in `MacroAssembler` which does this if `serialize` is not available.

https://www.felixcloutier.com/x86/serialize

Reply:
Right, we can throw `SERIALIZE` in the mix as well, at very least to match what `OrderAccess::cross_modify_fence` is doing. I believe it is cleaner to map it to another mode, rather than opting-in automatically. See new commit.